syntax = "proto3";

package post;

service PostService {
  rpc CreatePost (CreatePostRequest) returns (CreatePostResponse);
  rpc DeletePost (DeletePostRequest) returns (DeletePostResponse);
  rpc UpdatePost (UpdatePostRequest) returns (UpdatePostResponse);
  rpc GetPost (GetPostRequest) returns (GetPostResponse);
  rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);
  rpc ViewPost (ViewPostRequest) returns (ViewPostResponse);
  rpc LikePost (LikePostRequest) returns (LikePostResponse);
  rpc CommentPost (CommentPostRequest) returns (CommentPostResponse);
  rpc GetComments (GetCommentsRequest) returns (GetCommentsResponse);
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  string creator_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message CreatePostResponse {
  string post_id = 1;
  string created_at = 2;
}

message DeletePostRequest {
  string post_id = 1;
  string user_id = 2;
}

message DeletePostResponse {
  bool success = 1;
}

message UpdatePostRequest {
  string post_id = 1;
  string user_id = 2;
  string title = 3;
  string description = 4;
  bool is_private = 5;
  repeated string tags = 6;
}

message UpdatePostResponse {
  string updated_at = 1;
}

message GetPostRequest {
  string post_id = 1;
  string user_id = 2;
}

message GetPostResponse {
  Post post = 1;
}

message ListPostsRequest {
  string user_id = 1;
  int32 page = 2;
  int32 per_page = 3;
}

message ListPostsResponse {
  repeated Post posts = 1;
  int32 total = 2;
  int32 page = 3;
  int32 per_page = 4;
  int32 last_page = 5;
  int32 from_ = 6;
  int32 to_ = 7;
}

message Post {
  string post_id = 1;
  string title = 2;
  string description = 3;
  string creator_id = 4;
  string created_at = 5;
  string updated_at = 6;
  bool is_private = 7;
  repeated string tags = 8;
}

message ViewPostRequest {
  string post_id = 1;
  string user_id = 2;
}

message ViewPostResponse {
  bool success = 1;
}

message LikePostRequest {
  string post_id = 1;
  string user_id = 2;
}

message LikePostResponse {
  bool success = 1;
}

message CommentPostRequest {
  string post_id = 1;
  string user_id = 2;
  string comment = 3;
}

message CommentPostResponse {
  string comment_id = 1;
  string created_at = 2;
}

message GetCommentsRequest {
  string post_id = 1;
  string user_id = 2;
  int32 page = 3;
  int32 per_page = 4;
}

message GetCommentsResponse {
  repeated Comment comments = 1;
  Meta meta = 2;
}

message Comment {
  string comment_id = 1;
  string text = 2;
  string user_id = 3;
  string created_at = 4;
}

message Meta {
  int32 total = 1;
  int32 page = 2;
  int32 per_page = 3;
  int32 last_page = 4;
}